src/utils/Block.ts

```diff
@@ -0,0 +1,225 @@
+import { EventBus } from "./EventBus";
+import { nanoid } from 'nanoid';
+
+// Нельзя создавать экземпляр данного класса
+class Block<P extends Record<string, any> = any> {
+  static EVENTS = {
+    INIT: "init",
+    FLOW_CDM: "flow:component-did-mount",
+    FLOW_CDU: "flow:component-did-update",
+    FLOW_RENDER: "flow:render"
+  } as const;
+
+  public id = nanoid(6);
+  protected props: P;
+  public children: Record<string, Block>;
+  public childrenCollection: Array<any>;
+  private eventBus: () => EventBus;
+  private _element: HTMLElement | null = null;
+  private _meta: { tagName: string; props: P; };
+
+  /** JSDoc
+   * @param {string} tagName
+   * @param {Object} props
+   *
+   * @returns {void}
+   */
+  constructor(tagName = "div", propsWithChildren: P) {
+    const eventBus = new EventBus();
+
+    const { props, children } = this._getChildrenAndProps(propsWithChildren);
```

Comment:
И здесь еще подправим, чтобы можно было не указывать пропсы там, где они не нужны
```suggestion
  constructor(tagName = "div", propsWithChildren = {}) {
    const eventBus = new EventBus();

    const { props, children } = this._getChildrenAndProps(propsWithChildren as P);
```